### PR TITLE
Add Pootle FS app to manage syncing with filesystems

### PR DESCRIPTION
**Pull request by phlax**
This PR provides a pluggable system for configuring projects to sync to/from filesystems.

The configuration is managed using pootle_config and is stored against Projects.

Tracked Stores/files are remembered using a StoreFS object.

pootle_fs comes with a plugin for syncing to local fileystem

there is also a git plugin available here https://github.com/translate/pootle_fs_git

### CODE REVIEW

File: requirements/base.txt [hunk 1]
```diff
@@ -21,6 +21,7 @@ lxml>=2.2.0
 python-levenshtein
 rq>=0.5.0,<=0.5.6
 scandir
+dirsync
```
Comment (unho) [inline requirements/base.txt]: Sort alphabetically.

File: pootle/apps/pootle_fs/providers.py [hunk 2]
```diff
@@ -0,0 +1,18 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright (C) Pootle contributors.
+#
+# This file is a part of the Pootle project. It is distributed under the GPL3
+# or later license. See the LICENSE file for a copy of the license and the
+# AUTHORS file for copyright and authorship information.
+
+
```
Comment (unho) [inline pootle/apps/pootle_fs/providers.py]: Remove extra blank line.

File: pootle/apps/pootle_fs/response.py [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+FS_ACTION["pruned_from_fs"] = {
+    "title": "Pruned from filesystem",
+    "description":
+        ("Files removed from the filesystem that did not have matching Pootle "
+         "Stores")}
+FS_ACTION["fetched_from_fs"] = {
+    "title": "Fetched from filesystem",
+    "description":
+        ("Files staged from the filesystem that were new or newer than their "
+         "Pootle Stores")}
```
Comment (unho) [inline pootle/apps/pootle_fs/response.py]: According to description what is the difference with `pulled_to_pootle`?

File: pootle/apps/pootle_fs/response.py [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+FS_ACTION["added_from_pootle"] = {
+    "title": "Added from Pootle",
+    "description":
+        "Files staged from Pootle that were new or newer than their files"}
+FS_ACTION["pruned_from_fs"] = {
+    "title": "Pruned from filesystem",
+    "description":
+        ("Files removed from the filesystem that did not have matching Pootle "
+         "Stores")}
+FS_ACTION["fetched_from_fs"] = {
```
Comment (unho) [inline pootle/apps/pootle_fs/response.py]: What is the difference between staged and fetched?

File: pootle/apps/pootle_fs/response.py [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+        "Stores removed from Pootle that were not present in the filesystem"}
+FS_ACTION["pulled_to_pootle"] = {
+    "title": "Pulled to Pootle",
+    "description":
+        "Stores updated where filesystem version was new or newer"}
+FS_ACTION["added_from_pootle"] = {
+    "title": "Added from Pootle",
+    "description":
+        "Files staged from Pootle that were new or newer than their files"}
+FS_ACTION["pruned_from_fs"] = {
```
Comment (unho) [inline pootle/apps/pootle_fs/response.py]: I know this is an `OrderedDict` but is it no possible to put the `prune*` entries together?

File: pootle/apps/pootle_fs/response.py [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+FS_ACTION = OrderedDict()
+FS_ACTION["pruned_from_pootle"] = {
+    "title": "Deleted in Pootle",
+    "description":
+        "Stores removed from Pootle that were not present in the filesystem"}
+FS_ACTION["pulled_to_pootle"] = {
+    "title": "Pulled to Pootle",
+    "description":
+        "Stores updated where filesystem version was new or newer"}
+FS_ACTION["added_from_pootle"] = {
```
Comment (unho) [inline pootle/apps/pootle_fs/response.py]: Is this "pulled_to_fs"?

File: pootle/apps/pootle_fs/response.py [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+        ("Merged - FS won where unit updates conflicted")}
+FS_ACTION["merged_from_pootle"] = {
+    "title": "Merged from pootle",
+    "description":
+        ("Merged - Pootle won where unit updates conflicted")}
+
+
+class ProjectFSItemResponse(ItemResponse):
+
+    def __init__(self, *la, **kwa):
```
Comment (unho) [inline pootle/apps/pootle_fs/response.py]: Please don't forget about changing `*la, **kwa` to `*args, **kwargs`.

File: pootle/apps/pootle_fs/response.py (hunk 6 above)
Comment (phlax) [inline pootle/apps/pootle_fs/response.py]: no, on lots of levels

firstly it doesnt do the sync it just adds tracking, secondly anything going to the fs is referred to as pushed

File: pootle/apps/pootle_fs/response.py (hunk 7 above)
Comment (phlax) [inline pootle/apps/pootle_fs/response.py]: yep, k

File: pytest_pootle/env.py [hunk 8]
```diff
@@ -163,6 +178,91 @@ def setup_content_type(self):
 
         return content_type
 
+    def setup_fs(self):
+        from django.conf import settings
+
+        from .utils import update_store
+        from .fs.utils import create_test_suite
```
Comment (unho) [inline pytest_pootle/env.py]: Nit: `.fs` before `.utils`

File: pytest_pootle/env.py [hunk 9]
```diff
@@ -163,6 +178,91 @@ def setup_content_type(self):
...
+        import pytest_pootle
+
+        from .factories import ProjectDBFactory, TranslationProjectFactory
+
+        from pootle_app.models import Directory
+        from pootle_language.models import Language
+        from pootle_project.models import Project
+        from pootle_store.models import Store
+
+        from pootle_fs.plugin import FSPlugin
```
Comment (unho) [inline pytest_pootle/env.py]: Nit: Put this alongside the other imports a few lines above.

File: pootle/apps/pootle_fs/response.py (hunk 6 above)
Comment (unho) [inline pootle/apps/pootle_fs/response.py]: So this is staging?

File: pytest_pootle/fixtures/pootle_fs/plugin.py [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+    return add
+
+
+def _generate_rm_fixtures():
+    rm = OrderedDict()
+    for test, pootle_path, fs_path in FETCH_PATHS:
+        rm[test] = {
+            "pootle_path": pootle_path,
+            "fs_path": fs_path}
+    return rm
```
Comment (unho) [inline pytest_pootle/fixtures/pootle_fs/plugin.py]: What is the difference with `_generate_add_fixtures()`? This is basically repeated code.

Comment (unho) [inline pytest_pootle/fixtures/pootle_fs/plugin.py]: It is also the same as `_generate_fetch_fixtures`.

**Comment by unho** [comment]
The "Add providers for pootle_fs" commit can be squashed to another.

File: pootle/apps/pootle_fs/localfs.py [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+#
+# This file is a part of the Pootle project. It is distributed under the GPL3
+# or later license. See the LICENSE file for a copy of the license and the
+# AUTHORS file for copyright and authorship information.
+
+import logging
+
+import dirsync
+
+from pootle_fs.plugin import Plugin
```
Comment (unho) [inline pootle/apps/pootle_fs/localfs.py]:
```python
from .plugin import Plugin
```

File: pootle/apps/pootle_fs/plugin.py [hunk 12]
```diff
@@ -0,0 +1,344 @@
...
+    def response(self, state):
+        return self.response_class(state)
+
+    def state(self, fs_path=None, pootle_path=None):
+        """
+        Get a state object for showing current state of FS/Pootle
+
+        :return state: Where ``state`` is an instance of self.state_class
+        """
+        self.pull()
```
Comment (ta2-1) [inline pootle/apps/pootle_fs/plugin.py]: should we pull in any case?

Comment (phlax) [inline pootle/apps/pootle_fs/plugin.py]: yep, whenever we call  state we want the most up-to-date state we can

File: requirements/base.txt [hunk 13]
```diff
@@ -15,13 +15,15 @@ django-transaction-hooks
...
 diff-match-patch>=20121119
+dirsync
 elasticsearch>=1.0.0,<1.7
 jsonfield
 lxml>=2.2.0
 python-levenshtein
 rq>=0.5.0,<=0.5.6
 scandir
 
+
```
Comment (unho) [inline requirements/base.txt]: Unnecessary blank line.

File: pootle/apps/pootle_fs/plugin.py [hunk 14]
```diff
@@ -0,0 +1,344 @@
...
+        User = get_user_model()
+        username = config.get(
+            self.project.__class__,
+            instance=self.project,
+            key="pootle_fs.pootle_user")
+        if username:
+            try:
+                return User.objects.get(username=username)
+            except User.DoesNotExist:
+                logger.warning(
```
Comment (unho) [inline pootle/apps/pootle_fs/plugin.py]: Only logging a warning? I would expect the need to have a user to be a blocker.

File: pootle/apps/pootle_fs/plugin.py [hunk 15]
```diff
@@ -0,0 +1,344 @@
...
+                    "Misconfigured pootle_fs user: %s"
```
Comment (unho) [inline pootle/apps/pootle_fs/plugin.py]: This is missing user rather than misconfigured user, isn't it?

File: pootle/apps/pootle_fs/plugin.py [hunk 16]
```diff
@@ -0,0 +1,344 @@
...
+        return fs_resources.get(self.__class__)(self.project)
+
+    @cached_property
+    def state_class(self):
+        return pootle_state.get(self.__class__)
+
+    @lru_cache(maxsize=None)
+    def get_fs_path(self, pootle_path):
+        """
+        Reverse match an FS filepath from a ``Store`` using the project fsconf.
```
Comment (unho) [inline pootle/apps/pootle_fs/plugin.py]: What is fsconf?

File: pootle/apps/pootle_fs/plugin.py [hunk 17]
```diff
@@ -0,0 +1,344 @@
...
+
+    def find_translations(self, fs_path=None, pootle_path=None):
+        """
+        Find translation file from the file system
+
+        :param force: Add conflicting translations.
+        :param fs_path: Path glob to filter translations matching FS path
+        :param pootle_path: Path glob to filter translations to add matching
+          ``pootle_path``
+        :yields pootle_path, fs_path:
```
Comment (unho) [inline pootle/apps/pootle_fs/plugin.py]: Is the trailing `:` ok?

File: pootle/apps/pootle_fs/plugin.py [hunk 18]
```diff
@@ -0,0 +1,344 @@
...
+        :param force: Add conflicting translations.
+        :param fs_path: Path glob to filter translations matching FS path
+        :param pootle_path: Path glob to filter translations to add matching
+          ``pootle_path``
+        """
+        to_add = state["pootle_untracked"]
+        if force:
+            to_add = (
+                to_add
+                + state["conflict_untracked"]
```
Comment (unho) [inline pootle/apps/pootle_fs/plugin.py]: `+`? Which type is `state["conflict_untracked"]`? If boolean I suggest using `or` instead.

Comment (unho) [inline pootle/apps/pootle_fs/plugin.py]: Forget about that. I see that `to_add` is a tuple.

File: pootle/apps/pootle_fs/plugin.py [hunk 19]
```diff
@@ -0,0 +1,344 @@
...
+            if fs_state.state_type == "conflict_untracked":
+                fs_state.kwargs["store_fs"] = self.create_store_fs(
+                    store=Store.objects.get(pootle_path=fs_state.pootle_path),
+                    fs_path=fs_state.fs_path)
+            fs_state.store_fs.file.merge(pootle_wins)
+            action_type = (
+                "staged_for_merge_pootle"
+                if pootle_wins
+                else "staged_for_merge_fs")
+            response.add(action_type, fs_state=fs_state)
```
Comment (unho) [inline pootle/apps/pootle_fs/plugin.py]: Shorter:

```python
            if pootle_wins:
                response.add("staged_for_merge_pootle", fs_state=fs_state)
            else:
                response.add("staged_for_merge_fs", fs_state=fs_state)
````

File: pootle/apps/pootle_fs/plugin.py [hunk 20]
```diff
@@ -0,0 +1,344 @@
...
+
+    @responds_to_state
+    def merge(self, state, response,
+              fs_path=None, pootle_path=None, pootle_wins=False):
+        """
+        :param fs_path: Path glob to filter translations matching FS path
+        :param pootle_path: Path glob to filter translations to add matching
+          ``pootle_path``
+        """
+        for fs_state in state["conflict"] + state["conflict_untracked"]:
```
Comment (unho) [inline pootle/apps/pootle_fs/plugin.py]: I would enclose in tuple: `state["conflict"] + state["conflict_untracked"]`

File: pootle/apps/pootle_fs/plugin.py [hunk 21]
```diff
@@ -0,0 +1,344 @@
...
+            store_fs = fs_state.store_fs
+            store_fs.file.pull(
+                merge=True,
+                pootle_wins=pootle_wins,
+                user=self.pootle_user)
+            store_fs.file.push()
+            store_fs.file.on_sync()
+            response.add(
+                response_type,
+                fs_state=fs_state)
```
Comment (unho) [inline pootle/apps/pootle_fs/plugin.py]: Why span 3 lines? It fits in a single line.

File: pootle/apps/pootle_fs/plugin.py [hunk 22]
```diff
@@ -0,0 +1,344 @@
...
+            fs_state.store_fs.file.pull(user=self.pootle_user)
+            fs_state.store_fs.file.on_sync()
+            response.add("pulled_to_pootle", fs_state=fs_state)
+        return response
+
+    @responds_to_state
+    @emits_state(pre=fs_pre_push, post=fs_post_push)
+    def sync_push(self, state, response, fs_path=None, pootle_path=None):
+        pushable = state['pootle_staged'] + state['pootle_ahead']
+        for fs_state in pushable:
```
Comment (unho) [inline pootle/apps/pootle_fs/plugin.py]:
```python
        for fs_state in (state['pootle_staged'] + state['pootle_ahead']):
```

File: pootle/apps/pootle_fs/plugin.py [hunk 23]
```diff
@@ -0,0 +1,344 @@
...
+    def sync_pull(self, state, response, fs_path=None, pootle_path=None):
+        """
+        :param fs_path: Path glob to filter translations matching FS path
+        :param pootle_path: Path glob to filter translations to add matching
+          ``pootle_path``
+        """
+        for fs_state in (state['fs_staged'] + state['fs_ahead']):
+            fs_state.store_fs.file.pull(user=self.pootle_user)
+            fs_state.store_fs.file.on_sync()
+            response.add("pulled_to_pootle", fs_state=fs_state)
```
Comment (unho) [inline pootle/apps/pootle_fs/plugin.py]: Is there no call to `self.pull` as for `self.push` in next method?

File: pootle/apps/pootle_fs/plugin.py (hunk 15 above)
Comment (phlax) [inline pootle/apps/pootle_fs/plugin.py]: no, misconfigured user

File: pootle/apps/pootle_fs/plugin.py [hunk 24]
```diff
@@ -0,0 +1,344 @@
...
+            fs_state.store_fs.file.push()
+            response.add('pushed_to_fs', fs_state=fs_state)
+        self.push(response)
+        for fs_state in response.completed("pushed_to_fs"):
+            fs_state.store_fs.file.on_sync()
+        return response
+
+    @responds_to_state
+    def sync_rm(self, state, response, fs_path=None, pootle_path=None):
+        """
```
Comment (unho) [inline pootle/apps/pootle_fs/plugin.py]: Does this method remove files that need to be removed when syncing?

File: pootle/apps/pootle_fs/plugin.py [hunk 25]
```diff
@@ -0,0 +1,344 @@
...
+        for fs_state in (state['fs_staged'] + state['fs_ahead']):
+            fs_state.store_fs.file.pull(user=self.pootle_user)
+            fs_state.store_fs.file.on_sync()
+            response.add("pulled_to_pootle", fs_state=fs_state)
+        return response
+
+    @responds_to_state
+    @emits_state(pre=fs_pre_push, post=fs_post_push)
+    def sync_push(self, state, response, fs_path=None, pootle_path=None):
```
Comment (unho) [inline pootle/apps/pootle_fs/plugin.py]: Some docstring explaining what the method does would be welcome. The same for other methods in this class.

File: pootle/apps/pootle_fs/plugin.py (hunk 20 above)
Comment (phlax) [inline pootle/apps/pootle_fs/plugin.py]: that wouldnt make it a tuple, and there is no reason to

File: pootle/apps/pootle_fs/plugin.py (hunk 15 above)
Comment (unho) [inline pootle/apps/pootle_fs/plugin.py]: Like in configuration points to a wrong/nonexistent user?

File: pootle/apps/pootle_fs/plugin.py (hunk 20 above)
Comment (unho) [inline pootle/apps/pootle_fs/plugin.py]: I suggested this to align with the other methods, so code follows the same pattern.

File: pootle/apps/pootle_fs/plugin.py (hunk 14 above)
Comment (phlax) [inline pootle/apps/pootle_fs/plugin.py]: this is for misconfigured user - its a pretty corner case, so i think logging that it cant use the configured user is appropriate.

File: pootle/apps/pootle_fs/plugin.py (hunk 23 above)
Comment (phlax) [inline pootle/apps/pootle_fs/plugin.py]: no, they happen at different times. `pull` happens before getting state, `push` happens when changes have been applied.

File: pootle/apps/pootle_fs/plugin.py [hunk 26]
```diff
@@ -0,0 +1,344 @@
...
+        pushable = state['pootle_staged'] + state['pootle_ahead']
+        for fs_state in pushable:
+            fs_state.store_fs.file.push()
+            response.add('pushed_to_fs', fs_state=fs_state)
+        self.push(response)
```
Comment (phlax) [inline pootle/apps/pootle_fs/plugin.py]: i might move this to the sync method tho @unho - not sure atm tbh

File: pootle/apps/pootle_fs/plugin.py (hunk 15 above)
Comment (phlax) [inline pootle/apps/pootle_fs/plugin.py]: correct

File: pootle/apps/pootle_fs/plugin.py (hunk 26 above)
Comment (unho) [inline pootle/apps/pootle_fs/plugin.py]: I would say it will make the code easier to follow.

Comment (phlax) [inline pootle/apps/pootle_fs/plugin.py]: tbh its more about where plugins (git, mercurial etc) need to intersect, and how to properly handle failure

File: pootle/apps/pootle_fs/plugin.py [hunk 27]
```diff
@@ -0,0 +1,344 @@
...
+    def create_store_fs(self, fs_path=None, pootle_path=None, store=None):
+        return self.store_fs_class.objects.create(
+            project=self.project,
+            pootle_path=pootle_path,
+            path=fs_path,
+            store=store)
+
+    def find_translations(self, fs_path=None, pootle_path=None):
+        """
+        Find translation file from the file system
```
Comment (ta2-1) [inline pootle/apps/pootle_fs/plugin.py]: Find translation fileS...

File: pootle/apps/pootle_fs/plugin.py (hunk 26 above)
Comment (ta2-1) [inline pootle/apps/pootle_fs/plugin.py]: > i might move this to the sync method tho @unho - not sure atm tbh

are there cases when `push()` is unnecessary in `sync()`?

Comment (phlax) [inline pootle/apps/pootle_fs/plugin.py]: perhaps, but i think that would be for the push method to decide

**Comment by ta2-1** [comment]
lgtm

**Comment by unho** [comment]
lgtm

**Pull request merged**